Add a check for packed tensors for convolution solvers

amberhassaan opened the pull request:
Some solvers allow non-packed tensor inputs while others don't. This PR adds a check to the ones that don't support.

amberhassaan committed:
Add a check for packed tensors for convolution solvers

CAHEK7 commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -622,6 +622,10 @@ static bool IsApplicableBase(const ExecutionContext& ctx, const ProblemDescripti
 {
     if(!problem.Is2d())
         return false;
+    if(problem.HasNonPackedTensors())
+    {
+        return false;
+    }
```
As far as I know that solver works fine with non-packet tensors. @Kirpich30000 @Slimakanzer, am I correct?

Kirpich30000 replied:
Winograd supports arbitrary strides for N, C, H, K, R dimensions in number of elements (fp16 or fp32). W on both input and output, and S on filters should be dense.

Kirpich30000 replied:
Non dense W and R is possible, but likely will greatly affect performance. Is there real interest in functional (but not performance) support for this?

CAHEK7 replied:
It still can be faster than naive convolution.

amberhassaan replied:
adding back the check to WinoGrad solvers at @atamazov 's recommendation

CAHEK7 commented on src/solver/conv_wino_fury_RxS.cpp:
```diff
@@ -171,6 +171,10 @@ bool ConvWinoFuryRxS<Winodata, Winofilter>::IsApplicable(const ExecutionContext&
     if(!problem.Is2d())
         return false;
 
+    if(problem.HasNonPackedTensors())
+    {
+        return false;
+    }
```
This solver works fine with strides too.

amberhassaan replied:
adding back the check to WinoGrad solvers at @atamazov 's recommendation

CAHEK7 reviewed (changes requested):
Some of the solvers should not be disabled, but I don't know about the others.

Let's discuss shall we disable all the solvers in this commit and re-enable some of them later or we'll spend some time and disable only those which do not support strides.

I vote for the second option - disable only some of the solvers which do not support strides.

iq136boy commented:
> Some of the solvers should not be disabled, but I don't know about the others.
> 
> Let's discuss shall we disable all the solvers in this commit and re-enable some of them later or we'll spend some time and disable only those which do not support strides.
> 
> I vote for the second option - disable only some of the solvers which do not support strides.

@CAHEK7 Do we have a list of solvers that support non-packed tensors? I would recommend to disable solvers that we are not sure if they support non-packed tensors or not. And only enable the solvers if they passed the non-packed tensors tests.

iq136boy commented on src/solver/conv_direct_naive_conv_fwd.cpp:
```diff
@@ -53,6 +53,10 @@ bool ConvDirectNaiveConvFwd::IsApplicable(const ExecutionContext& ctx,
+    if(problem.HasNonPackedTensors())
+    {
+        return false;
+    }
```
The naive conv fwd solver should support non-packed tensors, right?

amberhassaan replied:
correct.

CAHEK7 commented:
For sure I know only those two Winograd solvers. I'm not familiar with the others, but probably someone else is more familiar and can confirm it quickly.

Otherwise the solvers can be verified: if no stride information passed into kernel - that for sure does not support non-packed tensors, else do a simple direct test. But this may require a lot of time and probably should be a separate task.

amberhassaan committed:
remove check from naive and ck solvers that support strides

amberhassaan commented:
@CAHEK7 , @iq136boy : I have removed the checks from  `CK 3D grouped convolution` and `naive convolution` solvers, but for the rest, I'm not sure of the support myself.

CAHEK7 reviewed (approved):
Probably we need to create some issue to check the other solvers can they be applicable for non-packed tensors or not and re-enable them.

CAHEK7 commented:
> @CAHEK7 , @iq136boy : I have removed the checks from `CK 3D grouped convolution` and `naive convolution` solvers, but for the rest, I'm not sure of the support myself.

Could you re-enable two winograd solvers as well?

amberhassaan committed:
remove check from wingrad solvers

amberhassaan commented:
@CAHEK7 , @iq136boy , @junliume , @JehandadKhan : https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2478 created to develop a test that tests every solver against non-packed inputs.

averinevg commented on src/include/miopen/fusion/utils.hpp:
```diff
@@ -84,6 +84,10 @@ inline bool WinoCommonIsApplicable(const FusionContext& context, const FusionDes
         return false;
     if(!conv_problem.IsFp32())
         return false;
+    if(conv_problem.HasNonPackedTensors())
+    {
+        return false;
+    }
```
[Recommendation] Curly braces here and in other places may be omitted. The code in this case will look better when you need to check multiple conditions.

amberhassaan replied:
I can remove them. My personal preference is to always have `{}` and have the starting `{` on the same line.

averinevg commented on src/solver/conv_asm_1x1u_bias_activ_fused.cpp:
```diff
@@ -243,6 +243,10 @@ bool ConvBiasActivAsm1x1U::IsApplicable(const FusionContext& context,
     const auto conv_problem = problem.GetConvProblem(0, conv::Direction::Forward);
     const auto conv_ctx     = context.GetConvContext(conv_problem);
 
+    if(conv_problem.HasNonPackedTensors())
+    {
+        return false;
+    }
```
This condition is checked in `ConvAsm1x1U::IsApplicable()`

averinevg commented on src/solver/conv_ck_igemm_fwd_bias_activ_fused.cpp:
```diff
@@ -411,15 +411,18 @@ bool ConvCKIgemmFwdBiasActivFused::IsApplicable(const FusionContext& ctx,
         return false;
     const auto conv_problem = fdesc_problem.GetConvProblem(0, conv::Direction::Forward);
 
-    if(conv_problem.IsTensorsCasted())
-        return false;
```
[Recommendation] It's better not to move code that hasn't changed (here and in other places too)

amberhassaan replied:
So I moved these to order the checks by a guess of cost. Expensive checks should come after cheaper checks.

averinevg commented on src/include/miopen/conv/problem_description.hpp:
```diff
@@ -367,6 +367,17 @@ struct ProblemDescription : ProblemDescriptionBase
+        return !(in.IsPacked() && weights.IsPacked() && out.IsPacked());
+    }
+
+    bool HasMixedDataTypes() const
+    {
+        return !(GetInDataType() == GetWeightsDataType() &&
+                 GetWeightsDataType() == GetOutDataType());
+    }
+
```
:+1:

atamazov commented on src/include/miopen/conv/problem_description.hpp:
```diff
@@ -367,6 +367,17 @@ struct ProblemDescription : ProblemDescriptionBase
     bool IsNCHWc_NCHWc() const;
     bool IsNCHWc_CHWNc() const;
 
+    bool HasNonPackedTensors() const
```
[Recommendation] Formally speaking `HasNonPackedTensors` is incorrect when only one tensor is non-packed. I would say that `HasAtLeastOneNonPackedTensor` is fully correct, but I am not going to recommend this ;)

From the other hand, when we say "non-packed convolution" we imply that at least one tensor is non-packed, right? So maybe simply
```suggestion
    bool NonPacked() const
```
is better. Up to you; this is just a matter of taste.

amberhassaan replied:
@atamazov : favoring the current name as it is more descriptive.

atamazov replied:
[resolved]

atamazov commented on src/include/miopen/fusion/utils.hpp:
```diff
@@ -84,6 +84,10 @@ inline bool WinoCommonIsApplicable(const FusionContext& context, const FusionDes
+    if(conv_problem.HasNonPackedTensors())
+    {
+        return false;
+    }
```
No `{ }` around single statement (`return false` in this case) everywhere, please.

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/packed-check-add

amberhassaan committed:
remove redundant/repeated checks

amberhassaan committed:
remove braces around if statements

amberhassaan committed:
fix formatting

atamazov commented on src/solver/conv_ck_igemm_fwd_bias_activ_fused.cpp:
```diff
@@ -411,15 +411,18 @@ bool ConvCKIgemmFwdBiasActivFused::IsApplicable(const FusionContext& ctx,
         return false;
     const auto conv_problem = fdesc_problem.GetConvProblem(0, conv::Direction::Forward);
 
-    if(conv_problem.IsTensorsCasted())
```
Please revert all code moves. Thanks.

amberhassaan replied:
done in latest commit.

atamazov commented:
@amberhassaan 

> @CAHEK7 Do we have a list of solvers that support non-packed tensors? I would recommend to disable solvers that we are not sure if they support non-packed tensors or not. And only enable the solvers if they passed the non-packed tensors tests.

Please assume that all old solvers do not support non-packed tensors. Even if Winograd seems supporting them, enabling the feature require investigations and testing.

junliume commented:
@amberhassaan could you resolve the last round of review opinions above? We will let CI run once and then merge. Thanks!

amberhassaan committed:
Merge remote-tracking branch 'origin/develop' into amber/packed-check-add

amberhassaan committed:
address comments. Revert reordered checks. Re-add check to WinoGrad solvers

amberhassaan commented:
> @amberhassaan could you resolve the last round of review opinions above? We will let CI run once and then merge. Thanks!

@junliume : I have addressed all the comments. Please merge if CI passes.

averinevg commented on src/solver/conv_asm_7x7c3h224w224k64u2v2p3q3f1.cpp:
```diff
@@ -65,17 +68,11 @@ bool ConvAsm7x7c3h224w224k64u2v2p3q3f1::IsApplicable(const ExecutionContext& ctx
     if(!(name == "gfx800" || name == "gfx802" || name == "gfx803" || name == "gfx804" ||
          name == "gfx900" || name == "gfx904" || name == "gfx906" || name == "gfx908"))
-    {
         return false;
-    }
```
Curly braces are needed here. There was some misunderstanding here, I meant that curly braces are not required where the condition is short enough and the returned statement occupies one line. My comment related primarily to the new code. I should have explained this in more detail.

amberhassaan replied:
OK, this is quite arbitrary and I find it hard to agree with. The language allows two choices: (1) no braces when the conditional has a single statement in the body and that has nothing to do with the length of the condition. That can be a source of bugs in absence of formatting, so I use `{}` all the time (choice #2). What you're asking is an arbitrary style choice and I find it inconsistent. Should we define how long is a "long condition"?

CAHEK7 replied:
I think it must be formalized and forced by clang-format.
If it's not forced by clang-format, we are free to do whatever we want - and most probably we want to argue what's the best codestyle ever.
@atamazov @averinevg please add appropriate check for it and we automatically will follow. I see no point to check codestyle on review.

averinevg replied:
I'll see if I can set up a clang format for this

amberhassaan replied:
I'm of the opinion that we should always use `{}` and modify `.clang-format` to put the starting `{` on the same line as `if` so that code height remains somewhat small.

On the general topic of style, I've opened up https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2482 for discussion on adopting Google Style Guide.

atamazov replied:
@CAHEK7 
> If it's not forced by clang-format, we are free to do whatever we want

NO. If there is no enforcing tool, then code review comes into play. That's one of the purposes of code reviews.

The reviewers are spending time, and **_good authors value this and learn from the reviews_**. And professional engineers are able to adapt his/her own tastes to whatever practices used in the communities (they joined).

@amberhassaan 
Is it fine that a small code fix you were asked to make here transformed to a separate ticket, discussion thread here and additional work for Evgenii?

CAHEK7 replied:
@atamazov NO for you NO)
The problem happened because in the first comment it was recommended to remove the brackets for a single statement conditions (which was done) and later for some of the cases it was recommended to return the brackets back if the condition is long enough. It is too fuzzy and it's a matter of taste, not a rule.

clang-format can handle curly braced case in one way or another. Probably it handles it not ideally, but in uniform and automated way, which simplify everything. So, it must be used, moreover, if the ideal rules can't be formalized and each reviewer may have his own flavor of ideal solution.
And as you've already said - we must adapt our tastes to make the life easier.

Just to mention, I'm talking about the case which can be handled by clang-format.
If it can't handle the case with includes (quotes vs brackets), so I don't ask for it, but @pfultz2 provided a reference how to properly fix it.
If it can't handle naming convention, I don't ask for it, I'm asking for some sort of small notes like "`WeUseCamelCase` for types and classes and `we_use_underscores` for variables". It would help a lot for newcomers or if the code is mess.
Since it's an opensource project and available for everyone, it would be nice to have such a document.

And as I've already said - we want to argue about the codestyle, it happens over and over again.

amberhassaan replied:
@atamazov : no, it's not fine that we are discussing style issues that would've been easily handled if we had a style document. We are creating unnecessary back and forth for minor matters. We need to have a `developer's guide` document in the repository that documents all the nuances of things like `Invoker & InvokerFactory` and how `in is out and out is in` in backward convolution and style issues like include files and variable/function naming etc. Hence the discussion in #2483

averinevg commented on src/solver/conv_asm_1x1u_stride2.cpp:
```diff
@@ -505,13 +507,9 @@ bool ConvAsm1x1UV2::IsApplicable(const ExecutionContext& ctx,
     const std::string name = ctx.GetStream().GetDeviceName();
     if(name.find("gfx8") == std::string::npos && name.find("gfx9") == std::string::npos)
-    {
         return false;
-    }
```
Ditto

averinevg commented on src/solver/conv_asm_dir_BwdWrW1x1.cpp:
```diff
@@ -493,13 +494,9 @@ bool ConvAsmBwdWrW1x1::IsApplicable(const ExecutionContext& ctx,
 
     const std::string name = ctx.GetStream().GetDeviceName();
     if(name.find("gfx8") == std::string::npos && name.find("gfx9") == std::string::npos)
-    {
         return false;
-    }
```
Ditto

averinevg commented on src/solver/conv_direct_naive_conv_bwd.cpp:
```diff
@@ -44,15 +44,14 @@ bool ConvDirectNaiveConvBwd::IsApplicable(const ExecutionContext& ctx,
     if(!ConvDirectNaiveConvIsApplicableByKernelType(ctx, problem))
         return false;
 
+    if(!problem.direction.IsBackwardData())
+        return false;
```
Existing code has been moved

amberhassaan replied:
I will revert it but the existing order of checks itself is something to discuss as it seems arbitrary and based on what the programmer wrote at the time or copy-pasted from somewhere else.

averinevg replied:
I agree that the current order of checking conditions may not be optimal, but here we need to do a little research and evaluate the expected benefit from this. It's better to do this in a separate pull request

CAHEK7 commented:
Let's keep all the curly braced in this PR like they are, add appropriate clang-format rule, apply it and forget about this problem forever. I see no point to spend any time to review codestyle if it can be checked and applied automatically.

JehandadKhan commented:
> Let's keep all the curly braced in this PR like they are, add appropriate clang-format rule, apply it and forget about this problem forever. I see no point to spend any time to review codestyle if it can be checked and applied automatically.

I agree with this, formatting issues should never be part of a PR discussion since these checks can be automated.

JehandadKhan commented:
@junliume  This PR can be merged.

atamazov reviewed (commented):
LGTM!
